pages/dictionary/index.tsx

```diff
@@ -24,25 +27,104 @@ const normalizeForSearch = (text: string) => {
-    const matchesTranslations = translations.filter(({ cz_translation, ua_translation }) => {
-      return normalizeForSearch(cz_translation).includes(searchText) || normalizeForSearch(ua_translation).includes(searchText);
+  const lastContainerRef = useRef<HTMLDivElement | null>(null);
+  const searchContainer = useRef<HTMLDivElement | null>(null);
+  const searchButton = useRef<HTMLButtonElement | null>(null);
+  const translationsContainerRef = useRef<HTMLDivElement | null>(null);
+
+  // scrolls to top whenever user type in search input
+  useEffect(() => {
+    if (!search.trim() || window.scrollY === 0) return;
+    window.scrollTo({
+      top: 0,
     });
+  }, [search]);
+
+  //loads  more items once viewport is close enough to last element
+  useEffect(() => {
+    const lastTranslation = lastContainerRef.current;
+
+    const options = {
+      rootMargin: '100px',
+      threshold: [0],
+    };
+    const observer = new IntersectionObserver(([entry]) => {
+      const { intersectionRatio } = entry;
+      if (intersectionRatio > 0 && intersectionRatio < 1) {
+        setMaxItems((prev) => prev + 20);
+      }
+    }, options);
+
+    lastTranslation && observer.observe(lastTranslation);
+
+    return () => {
+      lastTranslation && observer.unobserve(lastTranslation);
+    };
+  });
+
+  // track when search input becomes sticky to apply styles
+  useEffect(() => {
+    const element = searchContainer.current;
+    const observer = new IntersectionObserver(
+      ([e]) => {
+        if (e.intersectionRatio < 1) {
+          setIsSticky(true);
+        } else if (e.intersectionRatio >= 1) {
+          setIsSticky(false);
+        }
+      },
+      { threshold: [1], rootMargin: '-57px 0px 0px 0px' }
+    );
+
+    element && observer.observe(element);
 
-    return matchesCategoryTitle || matchesTranslations.length > 0;
+    return () => {
+      element && observer.unobserve(element);
+    };
+  });
+
+  // filters translation based on user search
+  const filterTranslations = () => {
+    const res = flattenTranslations?.reduce(
+      (acc, cur) => {
+        const searchText = normalizeForSearch(search);
+        // checks if phrase is already in filtered translations to avoid filter duplicates
+        if (cur.cz_translation in acc.visited) {
+          return acc;
+        }
+        if (normalizeForSearch(cur.cz_translation).includes(searchText) || normalizeForSearch(cur.ua_translation).includes(searchText)) {
+          acc.visited[cur.cz_translation] = true;
+          acc.filtered.push(cur);
+        }
+        return acc;
+      },
+      { visited: {} as { [key: string]: boolean }, filtered: [] as TranslationType[] }
+    );
+
+    res && setFilteredTranslations(res.filtered);
   };
 
+  useEffect(() => {
+    if (!search.trim()) return;
+    filterTranslations();
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [search]);
+
+  // flatten translations array to prepare it for filtering
+  useEffect(() => {
```

Comment:
This probably does not need to be in a useEffect, you can just compute it when you need or store using `useMemo`.

Reply:
translations won't change after initial render, that's why I flatten them on first render. But probably I am missing something.